File: packages/expo-sensors/android/src/main/java/expo/modules/sensors/modules/AccelerometerModule.java
// Copyright 2015-present 650 Industries. All rights reserved.

package expo.modules.sensors.modules;

import android.content.Context;
import android.hardware.SensorEvent;
import android.hardware.SensorManager;
import android.os.Bundle;

import expo.core.Promise;
import expo.core.interfaces.ExpoMethod;
import expo.interfaces.sensors.SensorService;
import expo.interfaces.sensors.services.AccelerometerService;

public class AccelerometerModule extends BaseSensorModule {
  public AccelerometerModule(Context reactContext) {
    super(reactContext);
  }

  @Override
  public String getName() {
    return "ExponentAccelerometer";
  }

  @Override
  public String getEventName() {
    return "accelerometerDidUpdate";
  }

  @Override
  protected SensorService getSensorService() {
    return getModuleRegistry().getModule(AccelerometerService.class);
  }

  protected Bundle eventToMap(SensorEvent sensorEvent) {
    Bundle map = new Bundle();
    map.putDouble("x", sensorEvent.values[0] / SensorManager.GRAVITY_EARTH);
    map.putDouble("y", sensorEvent.values[1] / SensorManager.GRAVITY_EARTH);
    map.putDouble("z", sensorEvent.values[2]/  SensorManager.GRAVITY_EARTH);
    return map;
  }

  @ExpoMethod
  public void startObserving(Promise promise) {
    super.startObserving();
    promise.resolve(null);
  }

  @ExpoMethod
  public void stopObserving(Promise promise) {
    super.stopObserving();
    promise.resolve(null);
  }

  @ExpoMethod
  public void setUpdateInterval(int updateInterval, Promise promise) {
    super.setUpdateInterval(updateInterval);
    promise.resolve(null);
  }
}

File: packages/expo-sensors/android/src/main/java/expo/modules/sensors/modules/AccelerometerModule.kt
// Copyright 2015-present 650 Industries. All rights reserved.
package expo.modules.sensors.modules

import android.hardware.Sensor
import android.hardware.SensorManager
import android.os.Bundle
import expo.modules.interfaces.sensors.services.AccelerometerServiceInterface
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.sensors.UseSensorProxy
import expo.modules.sensors.createSensorProxy

private const val EventName = "accelerometerDidUpdate"

class AccelerometerModule : Module() {
  private val sensorProxy by lazy {
    createSensorProxy<AccelerometerServiceInterface>(EventName) { sensorEvent ->
      Bundle().apply {
        putDouble("x", (sensorEvent.values[0] / SensorManager.GRAVITY_EARTH).toDouble())
        putDouble("y", (sensorEvent.values[1] / SensorManager.GRAVITY_EARTH).toDouble())
        putDouble("z", (sensorEvent.values[2] / SensorManager.GRAVITY_EARTH).toDouble())
      }
    }
  }

  override fun definition() = ModuleDefinition {
    Name("ExponentAccelerometer")

    UseSensorProxy(this@AccelerometerModule, Sensor.TYPE_ACCELEROMETER, EventName) { sensorProxy }
  }
}
